x/mongo/driver/topology/server.go

```diff
@@ -375,36 +418,82 @@ func (s *Server) update() {
 		}
 		s.subscriptionsClosed = true
 		s.subLock.Unlock()
-		if conn == nil || conn.nc == nil {
-			return
+
+		// We don't need to take s.heartbeatLock here because closeServer is called synchronously when the select checks
+		// below detect that the server is being closed, so we can be sure that the connection isn't being used.
+		if s.conn != nil {
+			_ = s.conn.close()
 		}
-		conn.nc.Close()
 	}
-	for {
+
+	var disconnectedWhileWaiting bool
+	waitUntilNextCheck := func() {
+		// Wait until heartbeatFrequency elapses, an application operation requests an immediate check, or the server
+		// is disconnecting.
 		select {
+		case <-heartbeatTicker.C:
+		case <-checkNow:
 		case <-done:
-			closeServer()
+			disconnectedWhileWaiting = true
 			return
-		default:
 		}
 
+		// Ensure we only return if minHeartbeatFrequency has elapsed or the server is disconnecting.
 		select {
-		case <-heartbeatTicker.C:
-		case <-checkNow:
+		case <-rateLimiter.C:
 		case <-done:
-			closeServer()
+			disconnectedWhileWaiting = true
 			return
 		}
+	}
 
+	for {
+		// Check if the server is disconnecting. If the disconnect happened between the next check and now, we'll read
+		// a value from the done channel. If it happened during waitUntilNextCheck, the value has already been read
+		// from the channel, so we check the disconnectedWhileWaiting flag.
+		var disconnecting bool
 		select {
-		case <-rateLimiter.C:
 		case <-done:
+			disconnecting = true
+		default:
+		}
+		if disconnecting || disconnectedWhileWaiting {
```

Comment:
What do you think about you closing the `done` channel so that all selects return? If that change was made then we could remove `disconnectedWhileWaiting`.

Reply:
I like this idea. I didn't make this change before because it changes the behavior of `Disconnect`, but it seems safe, so I've made the change in my latest commit.